mux/mux.go

```diff
@@ -3,104 +3,87 @@ package mux
 	"google.golang.org/grpc"
 )
 
 const defaultGracePeriod = 5 * time.Second
+const grpcRequestMarker = "application/grpc"
 
-// Serve starts a TCP listener and serves the registered protocol
-// servers (i.e., http, grpc, etc.) and blocks until server exits.
-// Context can be cancelled to perform graceful shutdown.
-// Use WithHTTP() and WithGRPC() to register procol servers.
+// Serve starts a TCP listener and serves the registered protocol servers and blocks
+// until server exits. Context can be cancelled to perform graceful shutdown.
+// Use WithHTTP() and WithGRPC() to register protocol servers.
 func Serve(ctx context.Context, listenAddr string, opts ...Option) error {
-	var mux cmuxWrapper
+	var mux protocolMuxServer
 	for _, opt := range opts {
 		if err := opt(&mux); err != nil {
 			return err
 		}
 	}
-	if mux.httpServer == nil && mux.grpcServer == nil {
+	if mux.httpHandler == nil && mux.grpcServer == nil {
 		return errors.New("at-least one of http & grpc server must be set")
 	}
-
 	return mux.Serve(ctx, listenAddr)
 }
 
-type cmuxWrapper struct {
-	httpServer  *http.Server
+type protocolMuxServer struct {
+	httpHandler http.Handler
 	grpcServer  *grpc.Server
 	gracePeriod time.Duration
 }
 
-func (cmw *cmuxWrapper) Serve(baseCtx context.Context, addr string) error {
+func (pmux *protocolMuxServer) Serve(baseCtx context.Context, addr string) error {
 	ctx, cancel := context.WithCancel(baseCtx)
 	defer cancel()
 
-	l, err := net.Listen("tcp", addr)
-	if err != nil {
-		return err
+	http1Server := &http.Server{
+		Addr:    addr,
+		Handler: h2c.NewHandler(pmux.muxedHandler(), &http2.Server{}),
 	}
-	defer l.Close()
-
-	m := cmux.New(l)
-	defer m.Close()
-
-	grpcL := m.MatchWithWriters(cmux.HTTP2MatchHeaderFieldSendSettings("content-type", "application/grpc"))
-	httpL := m.Match(cmux.Any())
-
-	errorCh := make(chan error, 3)
-	serveOnListener(httpL, cmw.httpServer, errorCh)
-	serveOnListener(grpcL, cmw.grpcServer, errorCh)
 
 	go func() {
-		if err := m.Serve(); err != nil {
-			errorCh <- err
+		err := http1Server.ListenAndServe()
+		if err != nil && !errors.Is(err, http.ErrServerClosed) {
+			log.Printf("[ERROR] server exited with error: %v", err)
+
+			// force-cancel the context so that graceful shutdown sequence exits as well.
+			cancel()
 		}
 	}()
 
-	select {
-	case <-ctx.Done():
-		return cmw.shutDown()
+	<-ctx.Done()
 
-	case e := <-errorCh:
-		_ = cmw.shutDown()
-		return e
-	}
-}
-
-func (cmw *cmuxWrapper) shutDown() error {
-	if cmw.grpcServer != nil {
-		cmw.grpcServer.GracefulStop()
-	}
-
-	if cmw.httpServer != nil {
-		shutdownCtx, cancel := context.WithTimeout(context.Background(), cmw.gracePeriod)
-		defer cancel()
+	// context has been cancelled (due to cancelled base-context or due to
+	// server exit).
+	shutdownCtx, cancel := context.WithTimeout(context.Background(), pmux.gracePeriod)
+	defer cancel()
 
-		return cmw.httpServer.Shutdown(shutdownCtx)
+	if pmux.grpcServer != nil {
+		pmux.grpcServer.GracefulStop()
```

Comment:
@spy16 When testing this graceful stop? did you by any chance experience this ?
https://github.com/grpc/grpc-go/issues/1384

Reply:
i know of this. but did not experience. the `httpServer` is being shut down before gRPC gracefulStop is called. so this issue does not happen. it performs grafceful shutdown as expected.